### ISSUE TITLE
Change main filename to run next

### ISSUE THREAD

**Issue by tannewt** [adafruit/circuitpython#1084]
Add the ability to set the next main filename before triggering a reload. This can be used to switch between games on a system or between memory heavy animation scripts. https://forums.adafruit.com/viewtopic.php?f=60&t=139041

**Comment by kevinjwalters** [comment]
My similar discussion: https://github.com/adafruit/circuitpython/issues/1084

**Comment by robertgallup** [comment]
Mentioned this awhile ago on Discord, but thought I'd chime in here.

I'm working on a tutorial and have a bunch of example scripts all stored on the CIRCUITPY drive. Switching between them requires manually renaming or copying files to code.py/main.py. This gets tedious so I wrote a menu script that lists the available other scripts on the CIRCUITPY drive. Selecting one runs it using Import. This works, in general. The problem is that larger scripts can overflow the heap when imported even though they work fine on their own.

There might be a more general solution, but for me, it would be great to have something like sys.exit(_'next.py'_) that would restart the vm with the named script. No need for this to disrupt any existing behavior. e.g. ctrl-D from the REPL and sys.exit() could still load code.py/main.py by default.

The way this would work for my case is that the menu would be code.py/main.py and use sys.exit('next.py') to run the chosen script. All the example scripts would act normally, you'd use ctrl-C,ctrl-D to restart the menu, no special hooks to get back to the menu. Of course, if you wanted to chain several scripts, each one could use sys.exit(_'next.py'_).

**Comment by deshipu** [comment]
I have exactly the same use case for my game consoles.

**Comment by dhalbert** [comment]
We would not enhance `sys.exit()` for that purpose, because we aim to be a subset or the same as the CPython version. But we could certainly add it to `supervisor`, such as adding an arg to [`supervisor.reload()`](https://circuitpython.readthedocs.io/en/latest/shared-bindings/supervisor/index.html?highlight=supervisor#supervisor.reload).

**Comment by robertgallup** [comment]
Adding it to supervisor.reload() would be perfect.

**Comment by tannewt** [comment]
Note that the hardest part of this is storing the filename because the heap gets reinitialized before running the file.

**Comment by robertgallup** [comment]
Good point. Is this same issue handled in `supervisor.set_next_stack_limit()`? Or, is it a different problem?

**Comment by tannewt** [comment]
Ya, it is similar but stack size is a fixed size so it's easy to just statically allocate it. We could simply have a char* array with a limit to start. If folks hit the limit often we could either increase it or make the allocation dynamic.

**Comment by robertgallup** [comment]
Ah ha. I see the issue. Picking a good limit is key (though, no matter what you pick, someone will have a use case for more :) ).

**Comment by dhalbert** [comment]
If the heap memory is not touched when the VM shuts down, then I think the following would work. It relies on being to able to use the stale data in the old heap.

1. Copy the location of the string object that is the filename into a pointer variable at a fixed location
2. Reload exception throws us back to somewhere in main.
3. VM gets shut down. Loop back to restart VM
4. Before heap is re-initialized (is it zeroed?), examine the string object pointed to. Allocate an array on the stack large enough to hold the filename, and copy the filename to there.
5. Initialize the heap.
6. Pass the filename on the stack to `run_code_py().

**Comment by deshipu** [comment]
I just realized there is one more piece of the puzzle missing — after the code finishes running, it would be best if the device restarted again, with the default `main.py` this time, instead of dropping into the REPL. I wonder if this is too much to ask for an obscure use case, though.

**Comment by dhalbert** [comment]
@deshipu If each of the alternate programs did a `supervisor.reload()` with no args when it's done, that would just happen. Are you thinking that is an extra piece of code that the alternative programs wouldn't have?

**Comment by deshipu** [comment]
Yes, ideally it would be the same kind of programs that you normally run as `code.py` — whatever the users have written or copied from examples. Requiring a reset at the end is a possible solution, but I'm sure it will often be missed and users will wonder why the device is not displaying the menu again after running their code.

**Comment by dhalbert** [comment]
[started to reply and then saw your edited reply]

It would be possible to set a flag that says to reload on exit, either via `supervisor.reload()` or some new function. But as I was going to mention, and you already did, most example programs already are infinite loops and have no exit mechanism, so they'd have to be modified anyway.

**Comment by deshipu** [comment]
Also, you probably don't want to restart if there was an error, only if it was a normal end of the program. I might be overthinking this.

**Comment by tannewt** [comment]
> Also, you probably don't want to restart if there was an error, only if it was a normal end of the program. I might be overthinking this.

I don't think you are overcomplicating it. I think it is complicated.

Another aspect is what file should be run on reload. We don't know what file has been changed when the FS is written. My guess is that we want the currently running file reloaded (and therefore we need to preserve the filename past start up into the new heap.)

A related use to this would be doing OTA file updates over wifi. We'll want a known good code.py to allow for writing new files even when the loaded file crashes.

**Comment by robertgallup** [comment]
So, I might be misunderstanding, but are you suggesting that `supervisor.reload('file_name')` would permanently override code.py/main.py as the file for reload? Or, are you talking about behavior on the next reload after an error? Or, neither of those...

**Comment by tannewt** [comment]
I think we'll want a separate function rather than putting it in `reload`. That way the code can set if for all reload causes, not just manual reload. Something closer to `set_next_stack_limit`: https://circuitpython.readthedocs.io/en/5.3.x/shared-bindings/supervisor/__init__.html#supervisor.set_next_stack_limit

Then we can have multiple kwargs for different reload reasons or separate functions. I imagine you want: reload, exception, safe mode and completion as reasons.

**Comment by robertgallup** [comment]
This sounds like a good general approach. As I envision my current use case, I just need to reload with a specific file without affecting any of the defaults. Though, with more options, my use case might evolve.

**Comment by deshipu** [comment]
After some thought, I think that just setting the file name for the next run would solve 90% of m case as well — the users always can reset the device to get back to the menu, and an "exit the game" option can be provided by the library, which would then do the correct thing.

**Comment by siddacious** [comment]
My loader for the OSHWA 2020 badge would appreciate this functionality. I assume that since the proposed mechanism is simply designating which file to run, there would be no need to make sure pins are released/de-initialized?

As you suggested previously @tannewt , this will mean we (I ) will need to handle the "don't reinit already initialized stuff" case for libraries. Since the libraries will need to know the 'commanded reload' vs 'default start' state, it would/could be exposed by`supervisor` or similar?

**Comment by tannewt** [comment]
@siddacious I was thinking it would be a full reload that would reset everything.

What is an example of things you don't want to reset?

**Comment by robertgallup** [comment]
For my use case, I think a full reload/reset would be optimal.

**Comment by siddacious** [comment]
@tannewt I was referring to our previous convo where you suggested thinking about how sensors libs should handle a restart, though in that context I think you were talking about waking up from a sleep that was entered for power saving purposes so I may have crossed the streams.

It sounds like this is a related but different issue as there are certainly use cases where you _would_ want a complete reset, sensors included ie: running an app that wants different configurations for a sensor.

**Comment by robertgallup** [comment]
One thing that just occurred to me is that, if possible, it might be useful to support `.mpy` files in this function. Currently, `.mpy ` isn't supported during a board restart.

**Comment by dhalbert** [comment]
@robertgallup I am not sure what you mean. Do you mean you can't have code.mpy?

If something.py and something.mpy both exist, then something.py will take precedence.

**Comment by deshipu** [comment]
I think I tested it a while ago, and indeed `code.mpy` or `main.mpy` weren't picked up.

**Comment by dhalbert** [comment]
We should fix that, but it's easy to have a code.py that imports the actual program.

**Comment by robertgallup** [comment]
@dhalbert @deshipu Yes. I haven't tested `code.mpy`, but `main.mpy` is not recognized. There was a discussion about this on Discord some time back.

**Comment by deshipu** [comment]
Sure, but it's always some extra memory, and the reason for using mpy is to save memory in the first place.

**Comment by robertgallup** [comment]
And the reason for this restart() function is to avoid using import to begin with.

**Comment by dhalbert** [comment]
I think it's just a matter of adding code.mpy and main.mpy to:

**Comment by robertgallup** [comment]
*That* would be a simple fix for the standard restart. :-) With this new function, you'd either have to accept the full file name, e.g. `cool_program.mpy' or make sure to check for the `.mpy` version (I guess the same way that `import` currently works.)

**Comment by tannewt** [comment]
I'd rather not support mpy at the top level because it obfuscates the running code and is impossible to edit.

**Comment by robertgallup** [comment]
A fair point. But, I'm curious, why is non-editable, obfuscated code undesirable? Are you worried about nefarious applications? Or, is it against the hack/education ethic of CircuitPython?

**Comment by tannewt** [comment]
It's against the open, accessible philosophy of CircuitPython.

The only advantage of it is reducing memory overhead during load. Otherwise, plain .py should be used.

Some folks think the obfuscation provides a level of secrecy of their code but that's more of an illusion than reality. All variable names are preserved and no optimization is done on the byte codes that I know of. So, it'd be straightforward getting something like the original code back. Only comments would be missing.

**Comment by robertgallup** [comment]
Agree. Having all code as `.py` would be optimal. But, since microcontrollers are resource constrained there are reasons for using `.mpy` files (e.g. the library bundle). And, once CircuitPython allows `.mpy` for modules, you can have a top level `main.py` that just imports a `.mpy` and it seems like you're back to code that's not open and accessible. So, not allowing `.mpy` at the top level doesn't seem to gain much and creates an artificial inconsistency. FWIW

**Comment by tannewt** [comment]
I'd like to move away from mpy in the long term. Allowing code.mpy will make that harder.

**Comment by robertgallup** [comment]
Makes sense. And, as you point out, there is very little real advantage for `.mpy` at the top level. And, it's not required for my use case.  :-)

**Comment by cwalther** [comment]
I have written a prototype implementation of what is being discussed here at [cwalther:setnextcode](https://github.com/cwalther/circuitpython/tree/setnextcode) (cwalther@f222a3f735f12f8fb630d700e408189cffc0669e). Not necessarily meant to be taken as-is (hence not a pull request yet), but just to have a basis for discussion. Maybe someone else is already working on it as well and we can combine our ideas. It covers the cases we need for pewpew-game/game-m4-menu#1, but maybe others have additional requirements. Reviews also appreciated because I am very unfamiliar with CircuitPython/MicroPython internals, so there may be better ways to do things.

**Comment by tannewt** [comment]
@cwalther Would you mind doing a draft PR? PRs are much easier to comment on.

**Comment by cwalther** [comment]
Done!

I also extened it to accept `None` or an empty string to reset to the default search order, which may be useful when a program that was itself started as an alternative next code wants to reboot into the standard code.py/main.py rather than itself using `supervisor.reload()` (such as https://github.com/python-ugame/circuitpython-stage/blob/9596a5904ed757e6fbffcf03e7aa77ae9ecf5223/pewpew_m4/ugame.py#L39), and may make it unnecessary to add a distinction between a reload from `supervisor.reload()` and one from USB filesystem writes.

**Comment by cwalther** [comment]
Based on @tannewt’s first comments on #3454, let’s collect some more detailed requirements.

My work so far has concentrated on the **game console** use case, which I envision as follows (@deshipu or @robertgallup, correct me if you have any additions):

* There is a menu, main.py (written by us experts), and a number of games: game.py, game2.py, … (written potentially by beginners).
* Turning on the device boots into the menu, which lets the user choose one of the games.
* Choosing “game.py” reboots into game.py.
* When game.py
  * is interrupted by a USB file write: reboot into game.py.
  * exits
    * successfully: immediately reboot into main.py.
    * with an exception: wait in the “Press any key to enter the REPL. Use CTRL-D to reload.” state so that the user can see the traceback.
      * When reloaded by USB file write, reboot into game.py. (This is not correctly handled by my current prototype, it reboots into main.py.)
      * When reloaded by CTRL-D immediately or CTRL-D in the REPL, reboot into either main.py or game.py, probably doesn’t matter which.
* Games should not need to do anything with `supervisor` or any other special action to exit “properly” and support the above workflow, they should just be able to run off the end or call `sys.exit()`.
* While developing a game, you can name it “code.py” to bypass the menu, afterwards you rename it to “game.py”.
* The same game.py should also work on a MicroPython device that does not have `supervisor` (with a different main.py and different implementations of the libraries it uses) in a more traditional run-by-import workflow.

What would the **OTA update** use case look like? Something like this?

* code.py receives an updated version of itself and saves it as new_code.py.
* When it determines that the download has been successful (perhaps with a CRC check), it reboots into new_code.py.
* When new_code.py
  * exits with an exception: reboot (immediately or with waiting?) into code.py. code.py should be able to retrieve what the exception was and tell the user that the update failed.
  * gets this far without an exception: renames itself to code.py and reboots into code.py.
* What if more modules than just code.py need to be updated?

Are there any other use cases that are fundamentally different from the above?

Apart from retrieving an exception from the previous run, which seems like a somewhat separate feature, I see two areas where the requirements of the different use cases may differ:

**A.** After next code exits without having called `supervisor.set_next_code_file()`, do we preserve the next code info (next reload goes into the same file again) or clear it (next reload goes into code.py/main.py)?

.  | Game Console | OTA
-- | ------------ | ---
when calling `supervisor.reload()` | either | either
when interrupted by a USB file write | preserve | preserve
when completed with an exception | either | clear
when completed successfully | clear | either (never happens in the above workflow)

**B.** After next code exits, do we wait in the “Press any key to enter the REPL. Use CTRL-D to reload.” state or reload immediately?

.  | Game Console | OTA
-- | ------------ | ---
when calling `supervisor.reload()` | reload | reload
when interrupted by a USB file write | reload | reload
when completed with an exception | wait | ?
when completed successfully | reload | either (never happens in the above workflow)

My current prototype has B partially configurable using the `reload_on_success` keyword argument, but not A. Can we fill in the question marks or fix mistakes in the tables above to figure out if and how much either of them needs to be configurable? Or do we just decide to make them maximally configurable anyway? So far it seems like we never need to distinguish between the first two rows and never need them configurable. The last two rows are less clear. (I would like to have B configurable in the “successfully” row even if the table indicates that it wouldn’t need to be, because the behavior required by the “game console” use case there is different from the behavior after code.py/main.py after a hard reboot, and some use case we haven’t envisioned yet might want it the same.)

**Comment by robertgallup** [comment]
Great requirements spec! A couple of things caught my attention.

1) What's the use case for the requirement to reboot into game.py on a USB file write (in the case A table, above) ? (I'm not sure this is necessary/desirable for my use case -- but, possibly because I'm overlooking something.)

2) Exception handling. For my case, if "game.py" is running and I want to stop it and return to the main menu, I would expect to use CTRL+C, CTRL+D. From my read, it sounds like you're proposing this could return to either game.py or the main menu? I think my use case would require returning to the main menu.

**Comment by deshipu** [comment]
> What's the use case for the requirement to reboot into game.py on a USB file write (in the case A table, above)? (I'm not sure this is necessary/desirable for my use case -- but, possibly because I'm overlooking something.)

If you are working on your game, and modifying the file, you want to see your changes taking effect immediately without having to select it from the menu each time.

**Comment by cwalther** [comment]
> For my case, if "game.py" is running and I want to stop it and return to the main menu, I would expect to use CTRL+C, CTRL+D. From my read, it sounds like you're proposing this could return to either game.py or the main menu? I think my use case would require returning to the main menu.

Good point, I’ve changed that.

**Comment by robertgallup** [comment]
> If you are working on your game, and modifying the file, you want to see your changes taking effect immediately without having to select it from the menu each time.

OK. I see. To me, this seems like a special case that might lead to unintended consequences/confusion in other applications and might be better handled by a developer disabling or bypassing the menu during development (e.g. renaming "game.py" to "code.py"). But, that's from my particular perspective.

**Comment by cwalther** [comment]
Can you think of a concrete example of a situation where it would be unintended or confusing?

I don’t have a strong opinion either way, I guess I tend toward @deshipu’s view because I’m working on his devices, and also @tannewt has said above
> My guess is that we want the currently running file reloaded

**Comment by robertgallup** [comment]
> Can you think of a concrete example of a situation where it would be unintended or confusing?

I might be misconstruing this, but let's say that I'm working on a suite of programs. I might be making a change to a file other than the one that's currently running. In this case, I'd rather re-run from the top. Or, I might be adding a new game and want to see if the menu picks it up properly. Again, I'd want to rerun the menu. Trying to infer developer intent from a USB write can be difficult.

If the case were changed to "if game.py is running and game.py is updated, then rerun game.py" it might have less potential for interference.

In any case, I don't think this breaks my use case, it's just an observation that this is an exception to the general rule that currently controls autorun ... and, exceptions like this make me nervous.  :-)

**Comment by cwalther** [comment]
Thanks, that helped me understand why you see it as an “exception” or “special case”. I was operating under the assumption that “the general rule that currently controls autorun” was equivalent to “always rerun the file that’s currently running”, because there was only one file that that could be, so it wouldn’t be an exception. But now I see that that’s not true: If you only have main.py and main.py is running and then you save a new code.py that didn’t previously exist, it will run code.py.

Food for thought. Any other opinions?

> If the case were changed to "if game.py is running and game.py is updated, then rerun game.py" it might have less potential for interference.

Agreed, but as far as I understand, that’s not possible, or at least not without a lot more code.

**Comment by deshipu** [comment]
You also would get surprising behavior when you update a file that is imported in your `game.py` but not run directly.

**Comment by deshipu** [comment]
The way I see it, the easiest to understand and explain behavior (apart from the current "always run `main.py` or `code.py`") would be:
- if a program finished running, or there was an error (including `KeyboardInterrupt`), go back to the default behavior of running the `main.py` or `code.py`, and, depending on the options, show REPL, exception message with a pause, or reset immediately.
- if there was a filesystem change, reload the currently running program.

**Comment by tannewt** [comment]
>     * with an exception: wait in the “Press any key to enter the REPL. Use CTRL-D to reload.” state so that the user can see the traceback.

I'm not sure you actually want to ever enter this state because in the game console case you may not be connected to a host computer. I suppose you could reset the device to get out of it though. An alternative would be to have another `.py` or `main.py` to handle exceptions, kind of like a 404 page. That way it can be branded and also read inputs like buttons to go back to the menu. (CircuitPython doesn't know about these.)

Let's not worry too much about the OTA case at this point. It has separate issues because we may want to assume not having access to the device. I think the game console model will be enough to experiment with. (main.py is the bootloader and is never replaced. the user code can never finish if it wants to update itself another way.)

**Comment by cwalther** [comment]
> I'm not sure you actually want to ever enter this state because in the game console case you may not be connected to a host computer.

But then, if you don’t happen to be connected to a host computer, you can’t see the traceback (on the display of the device) and don’t know what the bug in your game was. (The game console use case I have in mind is more aimed at (beginner) game developers than game players. They may be “playing”, not “developing” the game at this moment, but still be interested in identifying bugs when they happen.)

> An alternative would be to have another `.py` or `main.py` to handle exceptions, kind of like a 404 page. That way it can be branded and also read inputs like buttons to go back to the menu. (CircuitPython doesn't know about these.)

That’s an interesting idea. I’ll experiment with that.

---

In the meantime I have gone ahead and made it “maximally configurable” for prototype testing: added keyword arguments for the rows (rows 1 and 2 combined = “on reload”, row 3 = “on error”, row 4 = “on success”) of both tables (A = “sticky”, B = “reload”). [Documentation here.](https://github.com/cwalther/circuitpython/blob/c97078dd841798c3c1c8b45370cabb4656c9f8da/shared-bindings/supervisor/__init__.c#L113)

When it turns out that some of these options are never used, always used with the same value, or always used together, we can then remove or combine them. My current expectation is that `reload_on_success` will stay, `sticky_on_reload` (the controversial one from the last few posts) will stay if we can’t agree on a default value for it, and the rest can go.

So my question for interested parties is: Using these options, can you make it work the way you need for your use cases? And if yes, with what values?

My answer for the PewPew game console case is “yes, reload_on_success=True, sticky_on_reload=True (preferred, not absolutely requried)”.

---

I notice we are at 6.0.0-beta.0 now. What are the chances of getting this into 6.0? I have no particular preference, just to set expectations. There is one commit waiting at https://github.com/python-ugame/circuitpython-stage already (submodule `frozen/circuitpython-stage`) that should go into 6.0 in my opinion. Once this lands, there will be more.

**Comment by deshipu** [comment]
Another option would be for the menu program to check for exception and display it (or otherwise handle) in a user-friendly manner. This way we can use the console's button to scroll and dismiss the message, for example.

**Comment by cwalther** [comment]
Right, I also came to that conclusion when I was mulling it over under the shower. That actually seems the more elegant solution to me. The ability to retrieve the last exception is required either way.

So until someone convinces me otherwise, I think there is no need to have different next file names for different outcomes. The capabilities of the current prototype plus retrieving the last exception are sufficient.

**Comment by deshipu** [comment]
So all we need is an option that says what happens when the program finishes — either show the exception and drop to REPL (the default), or get back to the main program, possibly with a stored exception, and let that program decide what to do next.

**Comment by cwalther** [comment]
Translated into prototype-speak, that option would be a combined `reload_on_error = reload_on_success` with a default of `False`, while the rest would be fixed at `sticky_on_reload=True`, `sticky_on_error=False`, `sticky_on_success=False`, if I understand you correctly. Works for me.

However, for now, until we have established that the “retrieve last exception” feature is feasible and can land at the same time (which I intend to try, but no promise), I would prefer to keep `reload_on_error` and `reload_on_success` separate, because without that feature we need `reload_on_error=False` and `reload_on_success=True`.

**Comment by robertgallup** [comment]
If I'm reading this correctly, I think for my case, all of the `sticky_on_...` flags would be `False` which looks like the default? So, in my main menu, to jump to a menu choice, I would use:

```
supervisor.set_next_code_file(filename='example.py')
supervisor.reload()
```

Is that correct?

And, if I press the reset button or disconnect/reconnect the device, everything returns to default, running `code.py` or `main.py`?

**Comment by cwalther** [comment]
If you’re okay with staying in the “Code done running. Waiting for reload. / Press any key to enter the REPL. Use CTRL-D to reload.” wait state after example.py has finished, that is correct, except that the `filename` argument is positional-only (that is not deliberate, but I didn’t see any way of making it positional-or-keyword).

What I do is
```python
supervisor.set_next_code_file(selected + '.py', reload_on_success=True, sticky_on_reload=True)
supervisor.reload()
```

The `reload_on_success=True` makes it skip the wait state after finishing without error and go right back to the menu, and the `sticky_on_reload=True` is our difference of opinion on what a USB file write should do.

And yes, a hard reset returns everything to default.

**Comment by robertgallup** [comment]
With/without keyword are both fine with me. Does CircuitPython have a convention?

It definitely sounds like 'sticky' options are useful in your application. It may be when I get a better feeling for them in practice I'll find their utility, but for now I'm happy with the default.  :-)

**Comment by tannewt** [comment]
> With/without keyword are both fine with me. Does CircuitPython have a convention?

keywords are generally preferred. When you have multiple positional arguments of the same type I like to make them required.

**Comment by tannewt** [comment]
> When it turns out that some of these options are never used, always used with the same value, or always used together, we can then remove or combine them. My current expectation is that `reload_on_success` will stay, `sticky_on_reload` (the controversial one from the last few posts) will stay if we can’t agree on a default value for it, and the rest can go.

This is fine for a prototype here but I wouldn't want to merge it in the "maximally configurable" state because removing APIs is much harder than adding more. For the PR, please do the limited API needed for the game case.

> I notice we are at 6.0.0-beta.0 now. What are the chances of getting this into 6.0? I have no particular preference, just to set expectations. There is one commit waiting at https://github.com/python-ugame/circuitpython-stage already (submodule `frozen/circuitpython-stage`) that should go into 6.0 in my opinion. Once this lands, there will be more.

We only need to worry about 6.0.0 if we want to remove/rename APIs. Adding APIs can be done after 6.0.0. (Removing/renaming would have to wait for 7.0.0)

I think the PR is blocked on getting an adequate but minimal API set, implemented and documented.

**Comment by cwalther** [comment]
Agreed. The “maximally configurable” state was never meant to be merged as is, it’s just a prototype for people to play with to figure out what we need. I was making progress on the “retrieve last exception” front when the supervisor heap bug distracted me, so as far as I’m concerned, we’re closing in on that.

**Comment by tannewt** [comment]
@cwalther I also realized this approach here is very similar to what we'd need for setting the USB descriptor from boot.py. #1015

**Comment by cwalther** [comment]
I’m unfamiliar with USB and don’t know what a descriptor is, so I can’t comment in detail, but one thing to note is that the more pieces of information we save using this approach, the more carefully we need to think about their interaction. If one needs to be cleared and another one is newly set and a third one needs to be preserved, and not all of this is known at the same time, in what order do we do what so we don't leave holes and waste memory? I‘m actually working on planning that through between “next code“ and “traceback” right now.

**Comment by cwalther** [comment]
I think I have a generic solution that will work for any number of items (as long as they all fit on the stack at once). Implementing and testing it will probably take me a few days though.

**Comment by deshipu** [comment]
I can imagine that the memory exceptions are going to be particularly difficult to handle. especially when they happen with a small allocation and there is little memory left to process it.

**Comment by tannewt** [comment]
@cwalther Sorry to distract on this. A USB descriptor defines what type of a USB device something is such as keyboard vs audio. It's similar to this because we'd construct it in `boot.py` and save it as a byte string. USB is initialized after boot.py and then lives through all subsequent code.py. So, it's a similar case of save a byte string from the VM and then move it into supervisor memory to init TinyUSB with. It shouldn't add complexity because it won't change after `boot.py`.

**Comment by cwalther** [comment]
OK, just so I get a rough idea to keep in the back of my mind while I implement the generic system, to make sure it will be ready for it:
* Does the supervisor allocation containing the USB descriptor have to live forever (until poweroff/hard reset), or can it be freed after TinyUSB initialization (i.e. before the first code.py run)?
* How many bytes roughly is a typical USB descriptor?
* What happens when the Python functions that set up the USB descriptor in boot.py are called in code.py? Can that happen?

**Comment by tannewt** [comment]
> Does the supervisor allocation containing the USB descriptor have to live forever (until poweroff/hard reset), or can it be freed after TinyUSB initialization (i.e. before the first code.py run)?

I think it needs to live forever because the host could ask for it at any time.

> How many bytes roughly is a typical USB descriptor?
If I had to guess it's 500 - 1000.

> What happens when the Python functions that set up the USB descriptor in boot.py are called in code.py? Can that happen?
I would just have them raise an exception. But maybe it'd be ok if the USB is unconnected.

**Comment by cwalther** [comment]
In other news, taking a closer look at `supervisor_move_memory()`, I see that what it does with `terminal_tilegrid_tiles` is of the same sort (saving stuff from the GC heap in supervisor allocations) and should be handled by the same mechanism. It could potentially be too large to fit on the stack though. Maybe what it does with the rgbmatrix and sharpdisplay buffers too, although those don’t need their contents copied. A revised system that can deal with those cases is taking shape in my head. I may be able to replace these three ad-hoc “hybrid allocators” with a central solution.

**Comment by cwalther** [comment]
Quick status update: Iteration 3 of the generic movable allocation system, built on top of the existing supervisor heap, works (as far as tested with the terminal tilegrid, more thorough tests still to be done), but adds a fair bit of code. Now working on iteration 4 to see if redesigning the supervisor heap to intrinsically support movable allocations will achieve the same goals with less code.
I hope nobody is waiting for me, cause I’m having fun down here in this rabbit hole! 🙂

**Comment by tannewt** [comment]
Sounds good @cwalther. Don't lose sight of how much added code is too much. Running the CI will build all the boards and tell you if any are out of space.

**Comment by dglaude** [comment]
I am wondering how does this interact with deep-sleep.
Could a program on the MagTag go to sleep but prepare what program to run at next wakeup?
I am still searching for a way to save state from sleep to wakeup.

**Comment by cwalther** [comment]
In the current implementation, no, it’s all stored in RAM, which is powered down during deep sleep. Moving it to any kind of non-volatile memory (I’m not familiar with what the ESP32-S2 offers in that regard) would be a new requirement.

There will probably be other things to discuss though regarding interactions with deep sleep, as I discovered today while fixing merge conflicts.

**Comment by hierophect** [comment]
Resolved (as described) by #3454

**Comment by cwalther** [comment]
@deshipu or anyone interested in sample code: Now that the prerequisites (`supervisor.set_next_code_file()` and `supervisor.get_previous_traceback()`) have landed, here is my work in progress on a menu for the PewPew M4 that makes use of them: [cwalther/game-m4-menu:supervisor](https://github.com/cwalther/game-m4-menu/tree/supervisor) ([diff](https://github.com/pewpew-game/game-m4-menu/compare/master...cwalther:supervisor)). I am not currently working on this, so if anyone wants to take it and run with it, feel free.

The error dialog part is still mostly mockup. [Here is the program I’ve been testing it with.](https://gist.github.com/cwalther/3a4e79f99a62281d4b1680cea82ad2ca)

It‘s currently hampered by #5057.
